drivers/can/can_xlnx.c

```diff
@@ -0,0 +1,1329 @@
+	 * no more pending messages left (indicated by ISR[RXNEMP] = 0). If any
+	 * read operation causes an underflow error (ISR[RXUFLW]) in the RX FIFO,
+	 * break out of the reception handling loop.
+	 */
+	do {
+		/* Read the raw data from the RX FIFO */
+		idr  = sys_read32(base_addr + CAN_XLNX_RXFIFO_ID_OFFSET);
+		dlcr = sys_read32(base_addr + CAN_XLNX_RXFIFO_DLC_OFFSET);
+		dw1  = sys_read32(base_addr + CAN_XLNX_RXFIFO_DATA1_OFFSET);
+		dw2  = sys_read32(base_addr + CAN_XLNX_RXFIFO_DATA2_OFFSET);
+
+		/*
+		 * Clear the FIFO not empty flag before reading the next message
+		 * -> will be re-asserted if more messages are pending in the RX FIFO.
+		 */
+		sys_write32(CAN_XLNX_IRQ_RX_NOT_EMPTY, base_addr + CAN_XLNX_ICR_OFFSET);
+
+		/* Start with the simplest item: DLC */
+		rx_frame.dlc = (dlcr >> CAN_XLNX_FIFO_DLCR_DLC_OFFSET) &
+				CAN_XLNX_FIFO_DLCR_DLC_MASK;
+		/* RX timestamp if feature is enabled */
+#if defined(CONFIG_CAN_RX_TIMESTAMP)
+		rx_frame.timestamp = (dlcr & CAN_XLNX_FIFO_DLCR_RXT_MASK);
+#endif
+
+		/*
+		 * Extract ID based on standard / extended ID indication.
+		 * RTR indication depends on standard / extended format.
+		 */
+		if (idr & CAN_XLNX_FIFO_IDR_IDE) {
+			rx_frame.id_type = CAN_EXTENDED_IDENTIFIER;
+
+			/* Extended ID in IDH and IDL */
+			rx_frame.id  = (idr >> CAN_XLNX_FIFO_IDR_IDL_OFFSET) &
+				CAN_XLNX_FIFO_IDR_IDL_MASK;
+			rx_frame.id |= ((idr >> CAN_XLNX_FIFO_IDR_IDH_OFFSET) &
+					CAN_XLNX_FIFO_IDR_IDH_MASK) << 18;
+
+			/* RTR -> acquire from [RTR] */
+			if (idr & CAN_XLNX_FIFO_IDR_RTR) {
+				rx_frame.rtr = CAN_REMOTEREQUEST;
+			} else {
+				rx_frame.rtr = CAN_DATAFRAME;
+			}
+		} else {
+			rx_frame.id_type = CAN_STANDARD_IDENTIFIER;
+
+			/* Standard ID in IDH only */
+			rx_frame.id  = (idr >> CAN_XLNX_FIFO_IDR_IDH_OFFSET) &
+				CAN_XLNX_FIFO_IDR_IDH_MASK;
+
+			/* RTR -> acquire from [SRRRTR] */
+			if (idr & CAN_XLNX_FIFO_IDR_SRRRTR) {
+				rx_frame.rtr = CAN_REMOTEREQUEST;
+			} else {
+				rx_frame.rtr = CAN_DATAFRAME;
+			}
+		}
+
+		/* Frame payload */
+		rx_frame.data_32[0] = ntohl(dw1);
+		rx_frame.data_32[1] = ntohl(dw2);
+
+		/*
+		 * The frame has been read from the RX FIFO unconditionally,
+		 * now check if it matches and registered RX filter.
+		 */
+		if (dev_data->filter_usage_mask) {
+			for (i = 0; i < CAN_XLNX_NUM_ACCEPTANCE_FILTERS; i++) {
+				/*
+				 * Skip the current filter entry if:
+				 * - it is unused (corresp. bit not set in filter_usage_mask)
+				 * - the filter's ID/mask don't match the RX frame's ID.
+				 */
+				if (!(dev_data->filter_usage_mask & BIT(i))) {
+					continue;
+				}
+
+				if (!can_utils_filter_match(&rx_frame,
+							    &dev_data->rx_filters[i].filter)) {
+					continue;
+				}
+
+				dev_data->rx_filters[i].cb(dev, &rx_frame,
+							   dev_data->rx_filters[i].cb_arg);
+			}
+		} else {
+			LOG_DBG("%s discarded received frame with ID 0x%X as no "
+				"RX filters are registered", dev->name, rx_frame.id);
```

Comment:
Please leave this out as it will make it almost impossible to debug anything on a loaded CAN bus.